File: CS/ExtendedDataNavigator/Navigator/ExtendedDataNavigator.cs
using DevExpress.XtraEditors;
using DevExpress.XtraGrid;
using System;
using System.ComponentModel;
using System.Linq;
using System.Windows.Forms;

namespace VerticalDataNavigator.Navigator
{
    [System.ComponentModel.DesignerCategory("")]
    public class ExtendedDataNavigator : GridControlNavigator
    {
        private NavigatorOrientation _NavigatorOrientation;
        private NavigatorButtonsTextLocation _PrevTextLocation;

        public ExtendedDataNavigator(GridControl control) : base(control) {
            _PrevTextLocation = this.TextLocation;
            NavigatorOrientation = NavigatorOrientation.Horizontal;
            
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        [Browsable(false)]
        public NavigatorOrientation NavigatorOrientation
        {
            get { return _NavigatorOrientation; }
            set 
            { 
                _NavigatorOrientation = value;
                OnOrientationChanged();
                ViewInfo.CalcViewInfo(
                    null, 
                    Control.MouseButtons, 
                    PointToClient(Control.MousePosition), 
                    ClientRectangle);
            }
        }

        void OnOrientationChanged() {
            if(NavigatorOrientation == NavigatorOrientation.Vertical) {
                _PrevTextLocation = this.TextLocation;
                this.TextLocation = NavigatorButtonsTextLocation.None;
                return;
            }
            this.TextLocation = _PrevTextLocation;
        }

        protected override NavigatorButtonsBase CreateButtons()
        { return new ExtendedDataNavigatorButtons(this); }

    }

    public enum NavigatorOrientation { Horizontal = 0, Vertical = 1 }

}

File: VB/ExtendedDataNavigator/Navigator/ExtendedDataNavigator.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid
Imports System.ComponentModel
Imports System.Windows.Forms

Namespace VerticalDataNavigator.Navigator
    <System.ComponentModel.DesignerCategory("")> _
    Public Class ExtendedDataNavigator
        Inherits ControlNavigator

        Private _NavigatorOrientation As NavigatorOrientation

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(False)> _
        Public Property NavigatorOrientation() As NavigatorOrientation
            Get
                Return _NavigatorOrientation
            End Get
            Set(ByVal value As NavigatorOrientation)
                _NavigatorOrientation = value
                ViewInfo.CalcViewInfo(Nothing, Control.MouseButtons, PointToClient(Control.MousePosition), ClientRectangle)
            End Set
        End Property

        Public Sub New()
            NavigatorOrientation = NavigatorOrientation.Horizontal
        End Sub


        Public Sub New(ByVal control As GridControl)
            Me.New()
        NavigatableControl = control
        End Sub

        Protected Overrides Function CreateButtons() As NavigatorButtonsBase
            Return New ExtendedDataNavigatorButtons(Me)
        End Function

    End Class

    Public Enum NavigatorOrientation
        Horizontal = 0
        Vertical = 1
    End Enum

End Namespace
